### ISSUE TITLE
Inconsistent eigh gradients

### ISSUE THREAD

**Issue by maedoc** [tensorflow/tensorflow#25645]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Mojave
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: n/a
- TensorFlow installed from (source or binary): `conda install -c conda-forge tensorflow` in a Python 3.6 environment
- TensorFlow version (use command below): v1.10.0-rc1-19-g656e7a2b34 1.10.0
- Python version: Python 3.6.8 :: Anaconda, Inc.
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"

**Describe the current behavior**
I'm porting an optimization involving calls to `eigh` from the [`autograd`](https://github.com/HIPS/autograd) library to Tensorflow. When computing the gradient of a loss function involving a subset of eigenvectors, I see fairly large large differences (in terms of `norm(grad_np - grad_tf)`).

In my application, this causes the optimization to fail completely under TensorFlow while succeeding with the `autograd` library (using same optimizer, step size etc). 

**Describe the expected behavior**
Different numerical implementations of derivatives should be consistent.  Comparing all other operations in the AD graph between autograd and TF showed no problems, and the same would be expected for eigh.

**Code to reproduce the issue**
```python
import numpy as np
import tensorflow as tf

def loss_z_np(z, G, w, ne=3):
    from autograd import jacobian, grad
    def loss(z):
        import autograd.numpy as np
        d = 2*z - np.sum(w, axis=1)
        e, v = np.linalg.eigh(np.diag(d) + w)
        ssev = np.sum(np.square(e[-ne:] * v[:, -ne:]), axis=-1)
        return np.sum(np.dot(G, ssev))
    return grad(loss)(z)

def lsa_loss(z, G, w, ne=3):
    d = 2*z - tf.reduce_sum(w, 1)
    e, v = tf.linalg.eigh(tf.diag(d) + w)
    ssev = tf.reduce_sum(tf.square(e[-ne:] * v[:, -ne:]), 1)
    return tf.reduce_sum(tf.matmul(G, tf.expand_dims(ssev, 1)))

def loss_z_tf(z, G, w, ne=3):
    G, w = [tf.convert_to_tensor(_, dtype=tf.float32) for _ in (G, w)]
    z = tf.Variable(tf.convert_to_tensor(z, dtype=tf.float32))
    loss = lsa_loss(z, G, w, ne=ne)
    lg = tf.gradients(loss, z)
    model = tf.global_variables_initializer()
    trace = []
    with tf.Session() as session:
        session.run(model)
        lg_np = session.run(lg)
    return lg_np


m, n = 79, 164
norms = []
for _ in range(30):
    G = np.random.rand(m, n).astype('f')
    w = np.random.rand(n, n).astype('f')
    z = - np.random.rand(n).astype('f')
    lg_np = loss_z_np(z, G, w, ne=10)
    lg_tf = loss_z_tf(z, G, w, ne=10)
    norms.append(np.linalg.norm(lg_np - lg_tf))
print('average norm', sum(norms)/len(norms))
```
This prints an average of ~30~ 16.

**Other info / logs**

n/a

_edit_ to switch the NumPy code to float32; problem persists.

**Comment by jvishnuvardhan** [comment]
@maedoc I think it was due to incorrect dtype used in the code. The dtype of lg_np was float64 where as dtype of lg.tf was float32. Because of the mismatch in the dtype, the difference in precision level of each result affect the norm. I modified the code here.

import numpy as np
import tensorflow as tf

def loss_z_np(z, G, w, ne=3):
    from autograd import jacobian, grad
    def loss(z):
        import autograd.numpy as np
        d = 2*z - np.sum(w, axis=1)
        e, v = np.linalg.eigh(np.diag(d) + w)
        ssev = np.sum(np.square(e[-ne:] * v[:, -ne:]), axis=-1)
        return np.sum(np.dot(G, ssev))
    return grad(loss)(z)

def lsa_loss(z, G, w, ne=3):
    d = 2*z - tf.reduce_sum(w, 1)
    e, v = tf.linalg.eigh(tf.diag(d) + w)
    ssev = tf.reduce_sum(tf.square(e[-ne:] * v[:, -ne:]), 1)
    return tf.reduce_sum(tf.matmul(G, tf.expand_dims(ssev, 1)))

def loss_z_tf(z, G, w, ne=3):
    G, w = [tf.convert_to_tensor(_, dtype=tf.float64) for _ in (G, w)]
    z = tf.Variable(tf.convert_to_tensor(z, dtype=tf.float64))
    loss = lsa_loss(z, G, w, ne=ne)
    lg = tf.gradients(loss, z)
    model = tf.global_variables_initializer()
    trace = []
    with tf.Session() as session:
        session.run(model)
        lg_np = session.run(lg)
    return lg_np


m, n = 79, 164
norms = []
for _ in range(30):
    G = np.random.rand(m, n).astype(np.float64)
    np.array([1,], dtype=np.uint64)
    w = np.random.rand(n, n).astype(np.float64)
    z = - np.random.rand(n).astype(np.float64)
    lg_np = loss_z_np(z, G, w, ne=10)
    lg_tf = loss_z_tf(z, G, w, ne=10)
    norms.append(np.linalg.norm(lg_np - lg_tf))
print('average norm', sum(norms)/len(norms))

Please let me know what you think. If you are agree with the solution, please close the issue. Thanks!

**Comment by maedoc** [comment]
Thanks for your reply.  Indeed switching float64 does produce a smaller norm. However, I realized that comparing autograd and TF is not necessary to demonstrate the issue, here is a repro which compares TF's gradient to a finite difference, 
```python
import numpy as np
import tensorflow as tf

def loss_np(z, G, w, ne=3):
    e, v = np.linalg.eigh(np.diag(2*z - np.sum(w, axis=1)) + w)
    ssev = np.sum(np.square(e[-ne:] * v[:, -ne:]), axis=1)
    return np.sum(np.dot(G, ssev))

def fd_grad(z, G, w, ne=3, d=1e-3):
    f = lambda z: loss_np(z, G, w, ne=ne)
    d_i = lambda i: d * (np.r_[:z.size] == i)
    f_z = f(z)
    return np.array([(f(z + d_i(i)) - f_z) / d for i in np.r_[:z.size]])

def loss_tf(z, G, w, ne=3):
    e, v = tf.linalg.eigh(tf.diag(2*z - tf.reduce_sum(w, 1)) + w)
    ssev = tf.reduce_sum(tf.square(e[-ne:] * v[:, -ne:]), 1)
    return tf.reduce_sum(tf.matmul(G, tf.expand_dims(ssev, 1)))

def tf_grad(z, G, w, ne=3):
    z, G, w = [tf.convert_to_tensor(_, dtype=tf.float64) for _ in (z, G, w)]
    z = tf.Variable(z)
    grad = tf.gradients(loss_tf(z, G, w), z)
    model = tf.global_variables_initializer()
    with tf.Session() as session:
        session.run(model)
        grad = session.run(grad)
    return grad


m, n = 79, 164
G = np.random.rand(m, n).astype(np.float64)
w = np.random.rand(n, n).astype(np.float64)
z = - np.random.rand(n).astype(np.float64)

print(np.linalg.norm(fd_grad(z, G, w, d=1e-3) - tf_grad(z, G, w)))
```
which shows errors too large to be accounted for by machine precision, IMO.

I have noticed the errors in the gradient are consistently smaller when the matrix input to `tf.linalg.eigh` is positive definite.  Is there an undocumented assumption about the input?

**Comment by MonicaGu** [comment]
I have compared the TensorFlow gradient result with PyTorch gradient result. There is no big difference between them. So I don't think this is a TensorFlow bug. Maybe there is something wrong with the numpy calculation.

This is my code to compare:
```
import numpy as np
import tensorflow as tf
import torch
from torch.autograd import Variable as V

def loss_np(z, G, w, ne=3):
    e, v = np.linalg.eigh(np.diag(2*z - np.sum(w, axis=1)) + w)
    ssev = np.sum(np.square(e[-ne:] * v[:, -ne:]), axis=1)
    return np.sum(np.dot(G, ssev))

def fd_grad(z, G, w, ne=3, d=1e-3):
    f = lambda z: loss_np(z, G, w, ne=ne)
    d_i = lambda i: d * (np.r_[:z.size] == i)
    f_z = f(z)
    return np.array([(f(z + d_i(i)) - f_z) / d for i in np.r_[:z.size]])

def loss_tf(z, G, w, ne=3):
    e, v = tf.linalg.eigh(tf.linalg.diag(2*z - tf.reduce_sum(w, 1)) + w)
    ssev = tf.reduce_sum(tf.square(e[-ne:] * v[:, -ne:]), 1)
    return tf.reduce_sum(tf.matmul(G, tf.expand_dims(ssev, 1)))

def tf_grad(z, G, w, ne=3):
    z, G, w = [tf.convert_to_tensor(_, dtype=tf.float64) for _ in (z, G, w)]
    z = tf.Variable(z)
    with tf.GradientTape() as g:
        g.watch(z)
        out = loss_tf(z, G, w)
    return g.gradient(out, z).numpy()

def loss_pt(z, G, w, ne=3):
    e, v = torch.symeig(torch.diag(2*z - torch.sum(w, dim=1)) + w, eigenvectors=True, upper=False)
    ssev = torch.sum(torch.pow(e[-ne:] * v[:, -ne:], 2), dim=1)
    return torch.sum(torch.matmul(G, ssev.reshape((164, 1))))

def pt_grad(z, G, w, ne=3):
    z = V(torch.from_numpy(z), requires_grad=True)
    G = V(torch.from_numpy(G), requires_grad=True)
    w = V(torch.from_numpy(w), requires_grad=True)
    #out = loss_pt(z1, G1, w1)
    e, v = torch.symeig(torch.diag(2*z - torch.sum(w, dim=1)) + w, eigenvectors=True, upper=False)
    ssev = torch.sum(torch.pow(e[-ne:] * v[:, -ne:], 2), dim=1)
    out = torch.matmul(G, ssev.reshape((164, 1))).sum()
    out.backward()
    return np.array(z.grad)


m, n = 79, 164
G = np.random.rand(m, n).astype(np.float64)
w = np.random.rand(n, n).astype(np.float64)
z = - np.random.rand(n).astype(np.float64)

print(np.linalg.norm(fd_grad(z, G, w, d=1e-3) - tf_grad(z, G, w)))
print(np.linalg.norm(pt_grad(z, G, w) - tf_grad(z, G, w)))
```

The outputs are:
```
16.9599469169133
1.789755161380324e-08
```

**Comment by maedoc** [comment]
Thanks for taking a look.  I tried PyTorch as well, but didn't include here since this is the TF repo.  That both TF & PT produce similar results is hardly suprising if they use the same underlying cuSolver library? </guess>

In any case, I came across this because when I tried to optimize my loss function, autograd (based on numpy) had no problem and produced correct results, whereas TF & PT both failed to make progress.  Given the three are all gradient-based, the only explanation is that TF & PT eigh solvers produce wrong results for non-positive-definite matrices.
